# **Startup Dataset Preprocessing & Analysis**


## **Data Preprocessing:**

---

**1. Load & Understand the Data:**


In [418]:
# Import libraries
import pandas as pd

In [419]:
# Load dataset
file_path = "big_startup_secsees_dataset.csv"
df = pd.read_csv(file_path)
df.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list funding_total_usd  \
0                                              Media          10000000   
1  Application Platforms|Real Time|Social Network...            700000   
2                                  Apps|Games|Mobile           3406878   
3                                        Curated Web           2000000   
4                                           Software                 -   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  
0               1         NaN       2015-01-05      2015-01-05  
1               2  2014-09-04       2014-03-01      2014-10-14  
2               1         NaN       2014-01-30      2014-01-30  
3               1  2007-01-01       2008-03-19      2008-03-19  
4               1  2010-01-01       2014-07-24      2014-07-24

In [420]:
# Check dataset shape
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")

Rows: 66368, Columns: 14


In [421]:
# Data types and non-null counts
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   permalink          66368 non-null  object
 1   name               66367 non-null  object
 2   homepage_url       61310 non-null  object
 3   category_list      63220 non-null  object
 4   funding_total_usd  66368 non-null  object
 5   status             66368 non-null  object
 6   country_code       59410 non-null  object
 7   state_code         57821 non-null  object
 8   region             58338 non-null  object
 9   city               58340 non-null  object
 10  funding_rounds     66368 non-null  int64 
 11  founded_at         51147 non-null  object
 12  first_funding_at   66344 non-null  object
 13  last_funding_at    66368 non-null  object
dtypes: int64(1), object(13)
memory usage: 7.1+ MB
None


In [422]:
# Check missing values per column
print(df.isnull().sum())

permalink                0
name                     1
homepage_url          5058
category_list         3148
funding_total_usd        0
status                   0
country_code          6958
state_code            8547
region                8030
city                  8028
funding_rounds           0
founded_at           15221
first_funding_at        24
last_funding_at          0
dtype: int64


In [423]:
# Check for duplicate rows
print(df.duplicated().sum())

0


In [424]:
# Checking null values in 'name' column
print(df['name'].isnull().sum())

1


In [425]:
# Imputing missing values in 'name' column by inferring its value from "permalink" column
df.loc[df["name"].isna(), "name"] = (
    df.loc[df["name"].isna(), "permalink"]
    .str.split("/")
    .str[-1]
    .str.replace("-", " ")
    .str.title()
)


In [426]:
# Checking null values in 'name' column after imputation
print(df['name'].isnull().sum())

0


The `permalink` column contains the values of the `name` column within it. We used the `permalink` column to extract and fill missing `name` values.

In [427]:
# Checking null values in 'homepage_url' column 
print(df['homepage_url'].isnull().sum())

5058


In [428]:
# Fill missing homepage URLs with 'unknown'
df['homepage_url'].fillna('unknown', inplace=True)
df.head()

C:\Users\mtayy\AppData\Local\Temp\ipykernel_25928\4175626297.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['homepage_url'].fillna('unknown', inplace=True)


permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list funding_total_usd  \
0                                              Media          10000000   
1  Application Platforms|Real Time|Social Network...            700000   
2                                  Apps|Games|Mobile           3406878   
3                                        Curated Web           2000000   
4                                           Software                 -   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  
0               1         NaN       2015-01-05      2015-01-05  
1               2  2014-09-04       2014-03-01      2014-10-14  
2               1         NaN       2014-01-30      2014-01-30  
3               1  2007-01-01       2008-03-19      2008-03-19  
4               1  2010-01-01       2014-07-24      2014-07-24

In [429]:
# Checking null values in 'homepage_url' column after filling it with 'Unknown'
print(df['homepage_url'].isnull().sum())

0


There is no way to infer the missing values of 'homepage_url' column and also its not as much important column for analysis. Hence, we imputed the missing values with 'Unknown'.

In [430]:
# Category distribution
print(df['category_list'].value_counts())

category_list
Software                                                                                                                                                                                                                                    3995
Biotechnology                                                                                                                                                                                                                               3615
E-Commerce                                                                                                                                                                                                                                  1332
Mobile                                                                                                                                                                                                                                      1177
Clean Technology      

In [431]:
# Checking null values in 'category_list' column 
print(df['category_list'].isnull().sum())

3148


In [432]:
# Imputing 'Unknown' in empty cells of 'category_list' column
df['category_list'] = df['category_list'].fillna('Unknown')

In [433]:
# Checking null values in 'category_list' column 
print(df['category_list'].isnull().sum())

0


In [434]:
# Extract unique categories from category_list
unique_categories = (
    df["category_list"]
    .dropna()
    .str.split("|")      # split on '|'
    .explode()           # turn list into rows
    .str.strip()         # remove extra spaces
    .unique()            # get distinct categories
)

print(f"Unique categories: {len(unique_categories)}")
print(unique_categories)

Unique categories: 859
['Media' 'Application Platforms' 'Real Time' 'Social Network Media' 'Apps'
 'Games' 'Mobile' 'Curated Web' 'Software' 'Biotechnology' 'Analytics'
 'E-Commerce' 'Entertainment' 'Networking' 'Real Estate' 'Web Hosting'
 'Health and Wellness' 'Internet Marketing' 'Public Relations' 'Cable'
 'Distribution' 'Education' 'Search' 'Art' 'Marketplaces' 'Beauty'
 'Local Businesses' 'Restaurants' 'Cosmetics' 'Hospitality' 'Health Care'
 'Health Care Information Technology' 'Hospitals' 'Advertising'
 'Image Recognition' 'Sales and Marketing' 'Social Media'
 'Social Media Advertising' 'Coffee' 'Delivery' 'Subscription Service'
 'Mobile Commerce' 'Social Commerce' 'Enterprise Software' 'Mobility'
 'Batteries' 'Electronics' 'Energy' 'iOS' 'Music' 'Technology' 'Fashion'
 'EdTech' 'Social Travel' 'Personal Health' 'Sports' 'Audio'
 'Health Diagnostics' 'Medical Devices' 'Personalization'
 'Pharmaceuticals' 'Politics' 'Business Services'
 'Small and Medium Businesses' 'SaaS' 'Unkn

In [435]:

# Define semantic mapping from granular to broad categories
category_mapping = {
    # Technology & Software
    'Technology': ['Software', 'Enterprise Software', 'SaaS', 'Mobile', 'Apps', 'iOS', 'Android', 'iPhone',
                   'iPad', 'Web Development', 'Web Tools', 'Cloud Computing', 'Cloud Infrastructure',
                   'Cloud Management', 'Cloud Security', 'Developer Tools', 'Developer APIs', 'IT Management',
                   'Information Technology', 'Internet', 'Productivity Software', 'Operating Systems',
                   'IT and Cybersecurity', 'Internet Infrastructure', 'Mobile Infrastructure'],

    # E-Commerce & Retail
    'E-Commerce & Retail': ['E-Commerce', 'Retail', 'Online Shopping', 'Shopping', 'Marketplaces',
                            'E-Commerce Platforms', 'Retail Technology', 'Mobile Commerce', 'Social Commerce',
                            'Fashion', 'Cosmetics', 'Beauty', 'Jewelry', 'Shoes', 'Eyewear'],

    # Health & Biotechnology
    'Health & Biotechnology': ['Biotechnology', 'Health Care', 'Health and Wellness', 'Medical', 'Medical Devices',
                                'Pharmaceuticals', 'Health Diagnostics', 'Hospitals', 'Personal Health',
                                'Mobile Health', 'mHealth', 'Therapeutics', 'Clinical Trials', 'Doctors',
                                'Physicians', 'Health Care Information Technology', 'Dental', 'Life Sciences',
                                'Bioinformatics', 'Genetic Testing', 'Biopharm', 'Nutraceutical'],

    # Finance & FinTech
    'Finance & FinTech': ['FinTech', 'Finance', 'Financial Services', 'Banking', 'Payments', 'Credit Cards',
                          'Insurance', 'Venture Capital', 'Investment Management', 'Wealth Management',
                          'Personal Finance', 'Bitcoin', 'Cryptocurrency', 'Trading', 'Stock Exchanges'],

    # Media & Entertainment
    'Media & Entertainment': ['Media', 'Entertainment', 'Music', 'Video', 'Film', 'Television', 'Games',
                              'Video Games', 'Mobile Games', 'News', 'Publishing', 'Content', 'Digital Media',
                              'Video Streaming', 'Photography', 'Video on Demand', 'Broadcasting'],

    # Social & Communication
    'Social & Communication': ['Social Network Media', 'Social Media', 'Messaging', 'Chat', 'Video Chat',
                               'Social Media Platforms', 'Communities', 'Social Media Marketing',
                               'Social Media Management', 'Blogging Platforms', 'Forums'],

    # Education & EdTech
    'Education & EdTech': ['Education', 'EdTech', 'E-Learning', 'Online Education', 'Language Learning',
                           'Universities', 'Colleges', 'K-12 Education', 'Textbooks', 'Educational Games',
                           'Tutoring', 'Corporate Training', 'Teaching STEM Concepts'],

    # Energy & Clean Tech
    'Energy & Clean Tech': ['Clean Technology', 'Energy', 'Solar', 'Renewable Energies', 'Clean Energy',
                            'Energy Storage', 'Batteries', 'Energy Efficiency', 'Smart Grid', 'Wind',
                            'Biofuels', 'Oil and Gas', 'Oil & Gas', 'Natural Gas Uses'],

    # Real Estate & Construction
    'Real Estate & Construction': ['Real Estate', 'Commercial Real Estate', 'Property Management',
                                   'Construction', 'Home Automation', 'Home Decor', 'Interior Design',
                                   'Architecture', 'Building Products'],

    # Transportation & Automotive
    'Transportation & Automotive': ['Transportation', 'Automotive', 'Cars', 'Electric Vehicles',
                                    'Public Transportation', 'Ride Sharing', 'Taxis', 'Logistics',
                                    'Shipping', 'Drones', 'Aerospace'],

    # Food & Hospitality
    'Food & Hospitality': ['Restaurants', 'Food Processing', 'Hospitality', 'Coffee', 'Hotels', 'Travel',
                           'Tourism', 'Wine And Spirits', 'Cooking', 'Recipes', 'Groceries', 'Organic Food',
                           'Specialty Foods', 'Tea', 'Craft Beer'],

    # Marketing & Advertising
    'Marketing & Advertising': ['Advertising', 'Marketing Automation', 'Social Media Advertising',
                                'Digital Marketing', 'Sales and Marketing', 'Public Relations',
                                'Internet Marketing', 'Email Marketing', 'SEO', 'Ad Targeting'],

    # Analytics & Data
    'Analytics & Data': ['Analytics', 'Big Data', 'Big Data Analytics', 'Data Visualization',
                         'Business Intelligence', 'Predictive Analytics', 'Machine Learning',
                         'Artificial Intelligence', 'Data Mining', 'Natural Language Processing', 'Data Integration', 'Personal Data', 'Data Privacy'],

    # Manufacturing & Industrial
    'Manufacturing & Industrial': ['Manufacturing', 'Industrial', 'Industrial Automation', 'Robotics',
                                   '3D Printing', 'Hardware', 'Electronics', 'Semiconductors', 'Sensors',
                                   'Advanced Materials', 'Material Science', 'Chemicals'],

    # Security & Cybersecurity
    'Security & Cybersecurity': ['Security', 'Network Security', 'Cyber Security', 'Information Security',
                                 'Data Security', 'Mobile Security', 'Physical Security', 'Fraud Detection',
                                 'Enterprise Security', 'Cloud Security'],

    # Agriculture & Environment
    'Agriculture & Environment': ['Agriculture', 'Farming', 'Organic', 'Aquaculture', 'Environmental Innovation',
                                  'Sustainability', 'Green', 'Recycling', 'Water Purification', 'Water'],

    # Business Services
    'Business Services': ['Business Services', 'Consulting', 'Professional Services', 'Human Resources',
                          'Recruiting', 'CRM', 'Sales Automation', 'Project Management', 'Collaboration',
                          'Customer Service', 'Business Analytics', 'B2B', 'Accounting'],

    # Other
    'Other': ['Startups', 'Non Profit', 'Incubators', 'Angels', 'Venture Capital'],

    # Keep 'Unknown' values as it is
    'Unknown': ['Unknown']
}

# Create reverse mapping (category -> broad domain)
reverse_mapping = {}
for broad_cat, specific_cats in category_mapping.items():
    for spec_cat in specific_cats:
        reverse_mapping[spec_cat.lower()] = broad_cat

def map_to_broad_category(category_str):
    """
    Maps a category string (potentially pipe-separated) to a broad category.
    """

    categories = [cat.strip() for cat in str(category_str).split('|')]
    for cat in categories:
        cat_lower = cat.lower()
        if cat_lower in reverse_mapping:
            return reverse_mapping[cat_lower]
    for cat in categories:
        cat_lower = cat.lower()
        for key in reverse_mapping.keys():
            if key in cat_lower or cat_lower in key:
                return reverse_mapping[key]
    tech_keywords = ['tech', 'software', 'app', 'digital', 'online', 'web', 'internet',
                     'mobile', 'platform', 'cloud', 'network']
    first_cat_lower = categories[0].lower()
    if any(keyword in first_cat_lower for keyword in tech_keywords):
        return 'Technology'
    return 'Other'

df['category'] = df['category_list'].apply(map_to_broad_category)

# Display category distribution
print("Category Distribution:")
print(df['category'].value_counts())
print(f"\nTotal rows: {len(df)}")
print(f"Unknown categories: {(df['category'] == 'Unknown').sum()}")

Category Distribution:
category
Technology                     17224
Health & Biotechnology          8827
E-Commerce & Retail             4983
Media & Entertainment           4745
Unknown                         3148
Marketing & Advertising         3050
Business Services               2851
Analytics & Data                2834
Finance & FinTech               2682
Manufacturing & Industrial      2566
Other                           2232
Energy & Clean Tech             2071
Social & Communication          1946
Education & EdTech              1868
Food & Hospitality              1696
Transportation & Automotive     1227
Real Estate & Construction      1115
Security & Cybersecurity         967
Agriculture & Environment        336
Name: count, dtype: int64

Total rows: 66368
Unknown categories: 3148
category
Technology                     17224
Health & Biotechnology          8827
E-Commerce & Retail             4983
Media & Entertainment           4745
Unknown                         3148
M

In [436]:
# Checking null values in 'category_list' column 
print(df['category'].isna().sum())

0



In 'category_list' column , we had around '3148' empty values, so we filled them with 'Unknown'. Then count of all distinct values in this column was around '27K', which is a huge number.But it was coming because each value further comtains multiple categories/words in it concatenated by '|' or '+' symbol. So then, we found all unique categories by splitting up the values by the symbols '|' or '+', for which the count we got was 858.
As still its a huge number of categories and its impossible to do analysis with such number of categories. Therefore, we handled it by using the semantic mapping (map multiple similar sub domains into one broader category), which is the most reasonable and practical solution for it.

In [437]:
# Preview cleaned data
df.head(70)

permalink                    name  \
0                   /organization/-fame                   #fame   
1                /organization/-qounter                :Qounter   
2   /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                 /organization/0-6-com                 0-6.com   
4        /organization/004-technologies        004 Technologies   
..                                  ...                     ...   
65      /organization/1366-technologies       1366 Technologies   
66                /organization/139shop                 139shop   
67               /organization/13th-lab                13th Lab   
68              /organization/140-proof               140 Proof   
69                /organization/140fire                 140Fire   

                         homepage_url  \
0                  http://livfame.com   
1              http://www.qounter.com   
2                 http://oneofthem.jp   
3                  http://www.0-6.com   
4   http://004gmbh.de/en/004-interact   
..                                ...   
65            http://www.1366tech.com   
66             http://www.139shop.com   
67                 http://13thlab.com   
68                http://140proof.com   
69                 http://140fire.com   

                                        category_list funding_total_usd  \
0                                               Media          10000000   
1   Application Platforms|Real Time|Social Network...            700000   
2                                   Apps|Games|Mobile           3406878   
3                                         Curated Web           2000000   
4                                            Software                 -   
..                                                ...               ...   
65                                      Manufacturing          71450000   
66                                         E-Commerce                 -   
67  Augmented Reality|Computer Vision|Games|iPad|M...            700000   
68      Advertising|Big Data Analytics|Interest Graph           5500000   
69  Advertising|Entertainment|Real Time|Social Med...            500000   

       status country_code state_code                 region           city  \
0   operating          IND         16                 Mumbai         Mumbai   
1   operating          USA         DE             DE - Other  Delaware City   
2   operating          NaN        NaN                    NaN            NaN   
3   operating          CHN         22                Beijing        Beijing   
4   operating          USA         IL  Springfield, Illinois      Champaign   
..        ...          ...        ...                    ...            ...   
65  operating          USA         MA                 Boston        Bedford   
66  operating          CHN        NaN                    NaN            NaN   
67   acquired          SWE         26              Stockholm      Stockholm   
68  operating          USA         CA            SF Bay Area  San Francisco   
69  operating          USA         CA            Los Angeles   Santa Monica   

    funding_rounds  founded_at first_funding_at last_funding_at  \
0                1         NaN       2015-01-05      2015-01-05   
1                2  2014-09-04       2014-03-01      2014-10-14   
2                1         NaN       2014-01-30      2014-01-30   
3                1  2007-01-01       2008-03-19      2008-03-19   
4                1  2010-01-01       2014-07-24      2014-07-24   
..             ...         ...              ...             ...   
65               9  2007-01-01       2008-03-27      2015-04-09   
66               1         NaN       2007-02-01      2007-02-01   
67               1  2010-01-01       2012-06-15      2012-06-15   
68               2  2010-01-11       2009-07-01      2011-04-28   
69               1  2010-01-01       2010-05-05      2010-05-05   

                      category  
0        Media & Entertainment  
1       Soc

In [438]:
# Checking null values in 'country_code' column
print(df['country_code'].isna().sum())

6958


In [439]:
# Fill missing country codes with 'Unknown'
df['country_code'] = df['country_code'].fillna('Unknown')

In [440]:
# Imputing null values in 'country_code' column by 'Unknown'
print(df['country_code'].isna().sum())

0


The rows where 'Country_code' column has null values, also has null values in 'state_code', 'region' and 'city' columns. Hence there's no way for us to infer this column value by backtracking. Therefore, we filled it with 'Unknown'.

In [441]:
# Checking null values in 'state_code' column
print(df['state_code'].isna().sum())

8547


In [442]:
# A: Imputing null values in 'state_code' column by 'Unknown' where 'Country_code' is also 'Unknown'
df.loc[(df['state_code'].isna()) & (df['country_code'] == 'Unknown'), 'state_code'] = 'Unknown'

In [443]:
# Checking null values in 'state_code' column after filling those values where 'country_code' is also missing
print(df['state_code'].isna().sum())

1589


In [444]:
# B: Imputing remaining null values in 'state_code' column by the most frequent state code in its respective country
mode_state = df.groupby('country_code')['state_code'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else 'Unknown')
df.loc[df['state_code'].isna(), 'state_code'] = mode_state[df['state_code'].isna()]

In [445]:
# Checking null values in 'state_code' column after catering both cases
print(df['state_code'].isna().sum())

0


In [446]:
# Checking null values in 'region' column
print(df['region'].isna().sum())

8030


In [447]:
# A: Imputing null values in 'region' column by 'Unknown' where 'Country_code' is also 'Unknown'
df.loc[(df['region'].isna()) & (df['country_code'] == 'Unknown'), 'region'] = 'Unknown'

In [448]:
# Checking null values in 'region' column after filling those values where 'country_code' is also missing
print(df['region'].isna().sum())

1072


In [449]:
# B: Imputing remaining null values in 'region' column by the most frequent state code in its respective country
# If no mode exists, leave as NaN (do not fill with 'Unknown')
mode_state = df.groupby('state_code')['region'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else pd.NA)
df.loc[df['region'].isna(), 'region'] = mode_state[df['region'].isna()]

In [450]:
# Checking null values in 'region' column after filling those values with the mode of the region in its relevant state_Code value
print(df['region'].isna().sum())

14


In [451]:
# Now imputing null values in 'region' by inferring them from 'country' column
mode_state = df.groupby('country_code')['region'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else 'Unknown')
df.loc[df['region'].isna(), 'region'] = mode_state[df['region'].isna()]

In [452]:
# Checking null values in 'region' column after filling those values where 'country_code' is also missing
print(df['region'].isna().sum())

0


In [453]:
# Checking null values in 'city' column
print(df['city'].isna().sum())

8028


In [454]:
# A: Imputing null values in 'city' column by 'Unknown' where 'Country_code' is also 'Unknown'
df.loc[(df['city'].isna()) & (df['country_code'] == 'Unknown'), 'city'] = 'Unknown'

In [455]:
# Checking null values in 'region' column after filling those values where 'country_code' is also missing
print(df['city'].isna().sum())

1070


In [456]:
df[df['city'].isna()]

permalink  \
66                                 /organization/139shop   
96                                 /organization/1d4-pty   
163                                 /organization/24tidy   
174                                   /organization/2c2p   
228                                /organization/360shop   
...                                                  ...   
65941           /organization/zhijiang-jonway-automobile   
65944                                /organization/zhitu   
65957                             /organization/zhui-xin   
66264                                /organization/zubka   
66306  /organization/zurn-international-e-commerce-co...   

                             name                     homepage_url  \
66                        139shop           http://www.139shop.com   
96                        1d4 Pty  http://www.immortaloutdoors.com   
163                        24tidy                http://24tidy.com   
174                          2C2P              http://www.2c2p.com   
228                       360SHOP        http://www.360shop.com.cn   
...                           ...                              ...   
65941  Zhijiang Jonway Automobile        http://www.jonwayauto.com   
65944                       Zhitu           http://www.imzhitu.com   
65957                    Zhui Xin                   http://zhui.cn   
66264                       Zubka             http://www.zubka.com   
66306                        Zurn              http://www.zurn.com   

                                      category_list funding_total_usd  \
66                                       E-Commerce                 -   
96                         Geospatial|Maps|Outdoors             40000   
163                                        Services           1856561   
174    E-Commerce|Mobile Commerce|Payments|Software          10000000   
228                             Enterprise Software           1623640   
...                                             ...               ...   
65941                                    Automotive          29030000   
65944                                   Photography            240727   
65957                                      Software           1623640   
66264                                    Consulting  3919186.37690815   
66306                                  Social Media                 -   

          status country_code state_code   region city  funding_rounds  \
66     operating          CHN         22  Beijing  NaN               1   
96     operating          NOR          2   Sydney  NaN               1   
163    operating          CHN         22  Beijing  NaN               2   
174    operating          SGP    Unknown  Unknown  NaN               3   
228    operating          CHN         22  Beijing  NaN               1   
...          ...          ...        ...      ...  ...             ...   
65941   acquired          CHN          2   Sydney  NaN               1   
65944  operating          CHN         22  Beijing  NaN               2   
65957  operating          CHN         22  Beijing  NaN               1   
66264     closed          GBR         B9     Lyon  NaN               1   
66306  operating          CHN         22  Beijing  NaN               1   

       founded_at first_funding_at last_funding_at  \
66            NaN       2007-02-01      2007-02-01   
96            NaN       2012-04-11      2012-04-11   
163           NaN       2013-12-27      2014-07-14   
174    2003-01-01       2011-01-01      2015-04-27   
228           NaN       2013-03-01      2013-03-01   
...           ...              ...             ...   
65941         NaN       2010-07-08      2010-07-08   
65944         NaN       2013-01-01      2013-04-01   
65957         NaN       2013-04-01      2013-04-01   
66264  2006-01-01       2007-02-09      2007-02-09   
66306  2013-01-01       2013-12-01      2013-12-01   

                          category  
66             E-Commerce & Retai

In [457]:
# Now imputing null values in 'city' by inferring them from 'city' column
mode_state = df.groupby('region')['city'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else pd.NA)
df.loc[df['city'].isna(), 'city'] = mode_state[df['city'].isna()]

In [458]:
# Checking null values in 'region' column
print(df['city'].isna().sum())

0


In [459]:
# Checking null values in the dataset
df.isna().sum()

permalink                0
name                     0
homepage_url             0
category_list            0
funding_total_usd        0
status                   0
country_code             0
state_code               0
region                   0
city                     0
funding_rounds           0
founded_at           15221
first_funding_at        24
last_funding_at          0
category                 0
dtype: int64

In [460]:
df.dtypes

permalink            object
name                 object
homepage_url         object
category_list        object
funding_total_usd    object
status               object
country_code         object
state_code           object
region               object
city                 object
funding_rounds        int64
founded_at           object
first_funding_at     object
last_funding_at      object
category             object
dtype: object

In [ ]:
# Converting the data types of 'founded_at', 'first_funding_at' and 'last_funding_at columns into 'date' type

